### DISCUSSION TITLE
Provide a way to generate types from defined routes

### DISCUSSION THREAD

**Discussion by sheremet-va** [posva/unplugin-vue-router#18] [Ideas]
Is it possible to provide a way to generate these types with routes defined by hand? I mean something like this:

```ts
import { generateTypes } from 'unplugin-vue-router'
import { routes } from './src/routes'
await generateTypes(routes)
```

For people who don't use file based routing. We can run a small plugin that runs this script when `./src/routes` or its dependencies change.

**Comment by posva** [comment]
Maybe we could expose part of the lower level APIs like the router tree or simply adding/removing pages but you would still need to handle all the parsing yourself as there is only code to parse filepaths like `/users/[id].vue` not `/users/:id`. That parser is in vue-router package through the `createRouterMatcher()` (or similar).

I exposed the context at d8d1ea7 in `unplugin-vue-router` to allow using the engine outside of plugins. It has a `scanPages()` that does the work (naming could change though) and you can find the full example at `src/index.ts`

**Comment by Rolanddoda** [comment]
@sheremet-va that's brilliant! I had pretty much the same idea as described in #127 but I think having a low level API would be better!

> there is only code to parse filepaths like /users/[id].vue not /users/:id.

@posva how difficult it could be to parse filepaths with `:id`? Most probably I am wrong, but I guess the plugin gets `/users/[id].vue` and generates the route config e.g. `{ path: '/users/:id', component: () => import('/src/pages/users/[id].vue') }`. In the case of `generateTypes(routes)` the route config is already there so you wouldn't even have to read the folders/files naming to generate the route config.🤔

**Reply by posva** [reply]
The code to parse paths should already be in the repo if you want to explore it. Check how the route tree is created
